src/platform_impl/linux/x11/ime/context.rs

```diff
@@ -1,41 +1,239 @@
+) -> i32 {
+    let client_data = unsafe { &mut *(client_data as *mut ImeContextClientData) };
+
+    client_data.text.clear();
+    client_data.cursor_pos = 0;
+    client_data.is_composing = true;
+    client_data
+        .event_sender
+        .send((client_data.window, CompositionStart()))
+        .expect("failed to send composition start event");
+    debug!("preedit start callback");
+    -1
+}
+
+extern "C" fn preedit_done_callback(
+    _xim: ffi::XIM,
+    client_data: ffi::XPointer,
+    _call_data: ffi::XPointer,
+) {
+    let client_data = unsafe { &mut *(client_data as *mut ImeContextClientData) };
+
+    if !client_data.text.is_empty() {
+        client_data.text.clear();
+        client_data.cursor_pos = 0;
+        client_data
+            .event_sender
+            .send((
+                client_data.window,
+                CompositionUpdate(
+                    String::from_iter(client_data.text.clone()),
+                    client_data.cursor_pos,
+                ),
+            ))
+            .expect("failed to send composition update event");
+    }
+    client_data
+        .event_sender
+        .send((client_data.window, CompositionEnd()))
+        .expect("failed to send composition end event");
+    client_data.is_composing = false;
+}
+
+extern "C" fn preedit_draw_callback(
+    _xim: ffi::XIM,
+    client_data: ffi::XPointer,
+    call_data: ffi::XPointer,
+) {
+    let client_data = unsafe { &mut *(client_data as *mut ImeContextClientData) };
+    let call_data = unsafe { &mut *(call_data as *mut XIMPreeditDrawCallbackStruct) };
+    client_data.cursor_pos = call_data.caret as usize;
+
+    let chg_range =
+        call_data.chg_first as usize..(call_data.chg_first + call_data.chg_length) as usize;
+    if chg_range.start > client_data.text.len() || chg_range.end > client_data.text.len() {
+        warn!(
+            "invalid chg range: buffer length={}, but chg_first={} chg_lengthg={}",
+            client_data.text.len(),
+            call_data.chg_first,
+            call_data.chg_length
+        );
+        return;
+    }
+
+    // NULL indicate text deletion
+    let mut new_chars = if call_data.text.is_null() {
+        Vec::new()
+    } else {
+        let xim_text = unsafe { &mut *(call_data.text) };
+        if xim_text.encoding_is_wchar > 0 {
+            return;
+        }
+        let new_text = unsafe { CStr::from_ptr(xim_text.string.multi_byte) };
+
+        String::from(new_text.to_str().expect("Invalid UTF-8 String from IME"))
+            .chars()
+            .collect()
+    };
+    let mut old_text_tail = client_data.text.split_off(chg_range.end);
+    client_data.text.split_off(chg_range.start);
+    client_data.text.append(&mut new_chars);
+    client_data.text.append(&mut old_text_tail);
+
+    client_data
+        .event_sender
+        .send((
+            client_data.window,
+            CompositionUpdate(
+                String::from_iter(client_data.text.clone()),
```

Comment:
Instead of cloning the `Vec`, we can build a `String` using `client_data.text.iter().collect()` to avoid an extra allocation.

Reply:
Fixed

Reply:
I see you've fixed one instance, but there are two remaining instances of the `String::from_iter` usage.

Reply:
Good catch. I replaced 2 of them as well